cloud/scope/managed_control_plane.go

```diff
@@ -669,6 +671,174 @@ func (s *ManagedControlPlaneScope) getSpecFromActual(cluster *oke.Cluster) *infr
+				return err
+			}
+		} else {
+			s.OCIManagedControlPlane.SetAddonStatus(*addon.Name, s.getStatus(resp.Addon))
+			// addon present, update it
+			err = s.handleExistingAddon(ctx, okeCluster, resp.Addon, addon)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	// for addons which are present in the status object but not in the spec, the possibility
+	// is that user deleted it from the spec. Hence disable the addon
+	for k, _ := range s.OCIManagedControlPlane.Status.AddonStatus {
+		// present in status but not in spec
+		if getAddon(addonSpec, k) == nil {
+			err := s.handleDeletedAddon(ctx, okeCluster, k)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (s *ManagedControlPlaneScope) getStatus(addon oke.Addon) infrav2exp.AddonStatus {
+	// update status of the addon
+	status := infrav2exp.AddonStatus{
+		LifecycleState:            common.String(string(addon.LifecycleState)),
+		CurrentlyInstalledVersion: addon.CurrentInstalledVersion,
+	}
+	if addon.AddonError != nil {
+		status.AddonError = &infrav2exp.AddonError{
+			Status:  addon.AddonError.Status,
+			Code:    addon.AddonError.Code,
+			Message: addon.AddonError.Message,
+		}
+	}
+	return status
+}
+
+func (s *ManagedControlPlaneScope) handleExistingAddon(ctx context.Context, okeCluster *oke.Cluster, addon oke.Addon, addonInSpec infrav2exp.Addon) error {
+	// if the addon can be updated do so
+	// if the addon is already in updating state, or in failed state, do not update
+	s.Info(fmt.Sprintf("Reconciling addon %s with lifecycle state %s", *addon.Name, string(addon.LifecycleState)))
+	if !(addon.LifecycleState == oke.AddonLifecycleStateUpdating ||
+		addon.LifecycleState == oke.AddonLifecycleStateFailed) {
+		addonConfigurationsActual := getActualAddonConfigurations(addon.Configurations)
+		// if the version changed or the configuration changed, update the addon
+		// if the lifecycle state is needs attention, try to update
+		if addon.LifecycleState == oke.AddonLifecycleStateNeedsAttention ||
+			!reflect.DeepEqual(addonInSpec.Version, addon.Version) ||
+			!reflect.DeepEqual(addonConfigurationsActual, addonInSpec.Configurations) {
+			s.Info(fmt.Sprintf("Updating addon %s", *addon.Name))
+			_, err := s.ContainerEngineClient.UpdateAddon(ctx, oke.UpdateAddonRequest{
+				ClusterId: okeCluster.Id,
+				AddonName: addon.Name,
+				UpdateAddonDetails: oke.UpdateAddonDetails{
+					Version:        addonInSpec.Version,
+					Configurations: getAddonConfigurations(addonInSpec.Configurations),
+				},
+			})
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (s *ManagedControlPlaneScope) handleDeletedAddon(ctx context.Context, okeCluster *oke.Cluster, addonName string) error {
+	resp, err := s.ContainerEngineClient.GetAddon(ctx, oke.GetAddonRequest{
+		ClusterId: okeCluster.Id,
+		AddonName: common.String(addonName),
+	})
+	if err != nil {
+		if ociutil.IsNotFound(err) {
+			s.OCIManagedControlPlane.RemoveAddonStatus(addonName)
```

Comment:
Should we return here? won't this continue down the rest of the logic so the `resp` could be nil on line 787?

Reply:
yep, good point. added a unit test also for this.